mmpose/models/heads/hybrid_heads/vis_head.py

```diff
@@ -0,0 +1,228 @@
+
+        Overwrite the original keypoint_scores.
+        """
+        if isinstance(pose_preds, tuple):
+            pose_pred_instances, pose_pred_fields = pose_preds
+        else:
+            pose_pred_instances = pose_preds
+            pose_pred_fields = None
+
+        batch_vis_np = to_numpy(batch_vis, unzip=True)
+
+        assert len(pose_pred_instances) == len(batch_vis_np)
+        for index, _ in enumerate(pose_pred_instances):
+            pose_pred_instances[index].keypoint_scores = batch_vis_np[index]
+
+        return pose_pred_instances, pose_pred_fields
+
+    def predict(self,
+                feats: Tuple[Tensor],
+                batch_data_samples: OptSampleList,
+                test_cfg: ConfigType = {}) -> Predictions:
+        """Predict results from features.
+
+        Args:
+            feats (Tuple[Tensor] | List[Tuple[Tensor]]): The multi-stage
+                features (or multiple multi-stage features in TTA)
+            batch_data_samples (List[:obj:`PoseDataSample`]): The batch
+                data samples
+            test_cfg (dict): The runtime config for testing process. Defaults
+                to {}
+
+        Returns:
+            Union[InstanceList | Tuple[InstanceList | PixelDataList]]: If
+            posehead's ``test_cfg['output_heatmap']==True``, return both
+            pose and heatmap prediction; otherwise only return the pose
+            prediction.
+
+            The pose prediction is a list of ``InstanceData``, each contains
+            the following fields:
+
+                - keypoints (np.ndarray): predicted keypoint coordinates in
+                    shape (num_instances, K, D) where K is the keypoint number
+                    and D is the keypoint dimension
+                - keypoint_scores (np.ndarray): predicted keypoint scores in
+                    shape (num_instances, K)
+                - keypoint_visibility (np.ndarray): predicted keypoints
+                    visibility in shape (num_instances, K)
+
+            The heatmap prediction is a list of ``PixelData``, each contains
+            the following fields:
+
+                - heatmaps (Tensor): The predicted heatmaps in shape (K, h, w)
+        """
+        if test_cfg.get('flip_test', False):
+            # TTA: flip test -> feats = [orig, flipped]
+            assert isinstance(feats, list) and len(feats) == 2
+            flip_indices = batch_data_samples[0].metainfo['flip_indices']
+            _feats, _feats_flip = feats
+
+            _batch_vis = self.vis_forward(_feats)
+            _batch_vis_flip = flip_visibility(
+                self.vis_forward(_feats_flip), flip_indices=flip_indices)
+            batch_vis = (_batch_vis + _batch_vis_flip) * 0.5
+        else:
+            batch_vis = self.vis_forward(feats)  # (B, K, D)
+
+        batch_vis.unsqueeze_(dim=1)  # (B, N, K, D)
+
+        if not self.use_sigmoid:
+            batch_vis = torch.sigmoid(batch_vis)
```

Comment:
Why is `sigmoid` applied to `batch_vis` when `use_sigmoid` is `False`?

Reply:
I think the value of visibility prediction should be between 0 and 1? If `use_sigmoid` is `False`, there won't be sigmoid in `vis_forward`.

Reply:
If the sigmoid function is always applied to the head outputs, what is the role of the `use_sigmoid` argument?

Reply:
> Why is `sigmoid` applied to `batch_vis` when `use_sigmoid` is `False`?

@Ben-Louis When `use_sigmoid` is True, a `nn.Sigmoid()` will be in the `vis_head`, namely, the `batch_vis` is between 0~1. Otherwise, we need to do sigmoid manually when `use_sigmoid` is False.

Reply:
Using `nn.Sigmoid()` and `BCELoss` separately is not officially recommended by Pytorch, a better way is to use `BCEWithLogitsLoss` and do sigmoid manually when a 0-1 score is needed. So I suggest him to set a `use_sigmoid` to keep the flexity.

Reply:
@Tau-J, thanks for the clarification. This can indeed be puzzling, perhaps we could add some comments for clarity? Also, focusing on a single recommended practice could be beneficial. BTW, in MMDetection, this parameter is generally integrated in loss modules like `mmdet.CrossEntropyLoss`. We might consider refining the `BCELoss` module with this parameter, rather than introducing a completely new loss module.